Userland/Libraries/LibSQL/Value.cpp

```diff
@@ -12,38 +12,112 @@
 #include <LibSQL/Serializer.h>
 #include <LibSQL/TupleDescriptor.h>
 #include <LibSQL/Value.h>
-#include <math.h>
 #include <string.h>
 
 namespace SQL {
 
-Value::Value(SQLType type)
-    : m_type(type)
+// We use the upper 4 bits of the encoded type to store extra information about the type. This
+// includes if the value is null, and the encoded size of any integer type. Of course, this encoding
+// only works if the SQL type itself fits in the lower 4 bits.
+enum class SQLTypeWithCount {
+#undef __ENUMERATE_SQL_TYPE
+#define __ENUMERATE_SQL_TYPE(name, type) type,
+    ENUMERATE_SQL_TYPES(__ENUMERATE_SQL_TYPE)
+#undef __ENUMERATE_SQL_TYPE
+        Count,
+};
+
+static_assert(to_underlying(SQLTypeWithCount::Count) <= 0x0f, "Too many SQL types for current encoding");
+
+enum class TypeData : u8 {
+    Null = 1 << 4,
+    Int8 = 2 << 4,
+    Int16 = 3 << 4,
+    Int32 = 4 << 4,
+    Int64 = 5 << 4,
+    Uint8 = 6 << 4,
+    Uint16 = 7 << 4,
+    Uint32 = 8 << 4,
+    Uint64 = 9 << 4,
+};
+
+template<typename Callback>
+static decltype(auto) downsize_integer(Integer auto value, Callback&& callback)
+{
+    if constexpr (IsSigned<decltype(value)>) {
+        if (AK::is_within_range<i8>(value))
```

Comment:
Do we need to use `AK::` here?

Reply:
We don't export `is_within_range` from Checked.h and I didn't want to touch an AK header to do that :upside_down_face:

Reply:
That's a perfectly reasonable excuse